extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java

```diff
@@ -1,16 +1,66 @@
 package io.quarkus.jackson.deployment;
 
+import java.util.Collection;
+
+import javax.inject.Inject;
+
+import org.jboss.jandex.*;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.substrate.ReflectiveHierarchyBuildItem;
 
 public class JacksonProcessor {
 
+    private static final DotName JSON_DESERIALIZE = DotName.createSimple(JsonDeserialize.class.getName());
+    private static final DotName BUILDER_VOID = DotName.createSimple(Void.class.getName());
+
+    @Inject
+    BuildProducer<ReflectiveClassBuildItem> reflectiveClass;
+
+    @Inject
+    BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchyClass;
+
+    @Inject
+    CombinedIndexBuildItem combinedIndexBuildItem;
+
     @BuildStep
-    void register(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
-        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false,
+    void register() {
+        addReflectiveClass(true, false,
                 "com.fasterxml.jackson.module.jaxb.JaxbAnnotationIntrospector",
-                "com.fasterxml.jackson.databind.ser.std.SqlDateSerializer"));
+                "com.fasterxml.jackson.databind.ser.std.SqlDateSerializer");
+
+        IndexView index = combinedIndexBuildItem.getIndex();
+
+        Collection<AnnotationInstance> pojoBuilderInstances = index.getAnnotations(JSON_DESERIALIZE);
+        for (AnnotationInstance pojoBuilderInstance : pojoBuilderInstances) {
+            if (AnnotationTarget.Kind.CLASS.equals(pojoBuilderInstance.target().kind())) {
+                addReflectiveHierarchyClass(pojoBuilderInstance.target().asClass());
+
+                AnnotationValue annotationValue = pojoBuilderInstance.value("builder");
+                if (null != annotationValue && AnnotationValue.Kind.CLASS.equals(annotationValue.kind())) {
+                    DotName builderClassName = annotationValue.asClass().name();
+                    if (!BUILDER_VOID.equals(builderClassName)) {
+                        ClassInfo builderClass = combinedIndexBuildItem.getIndex().getClassByName(builderClassName);
+                        addReflectiveHierarchyClass(builderClass);
+                    }
+                }
+            }
+        }
     }
 
+    private void addReflectiveHierarchyClass(ClassInfo classInfo) {
+        if (AnnotationTarget.Kind.CLASS.equals(classInfo.kind())) {
```

Comment:
I don't think this check is needed, it should also by of `Class` Kind.

Reply:
Then when that is removed, I think the `ClassInfo builderClass = combinedIndexBuildItem.getIndex().getClassByName(builderClassName);` which is used before this method is called can also be simplified